### PR TITLE
DCOS-61253 - consume mesos state stream as protobuf

### PR DESCRIPTION
**Pull request by pierrebeitz**
TL;DR

* savings in Mesos (cause it does not need to serialize to JSON)
* saving 40% (SOAK) to 70% (small cluster) bandwidth. might be worse for clusters with huge payloads (e.g. labels).
* saving ~15% in parsing time (on SOAK 2.0.0)
* A service's tasks show up after ~6 seconds now instead of ~16 seconds (on SOAK
from HH-office, 5 quite consistent samples each) after initial app load.

-------------------

using protobuf seems like a good idea, as mesos does not need to serialize its
state to json and does not need to send as much data over the wire.

on the our side it looks like parsing protobut is faster than parsing JSON -
apart from the savings on bandwidth.

we're evading @dcos/recordio, @dcos/mesos-client, @dcos/connection-manager,
@dcos/copychars and @dcos/http-service. :tada:

This implementation should work around the memory leak that copychars solved.
Also it evades the need to flush the XHR-connection by using a StreamReader on
the fetch-response.

We also fix an issue here that formerly caused the subscribe-event to be fetched
twice (resulting in using double the initial bandwidth).

A services tasks show up after ~6 seconds now instead of ~16 seconds (on SOAK
from HH-office).

on a cluster with just minimal load (added 6 tasks) we save ~70% bandwidth
(5.7kb vs 17.9kb) and save some time while parsing:
10000x protobuf   => 3635ms
10000x JSON.parse => 4169ms

on SOAK we save ~40% in bandwidth (6.2MB => 3.7MB) and abaut the same percentage
in parsing as above.

Tradeoffs
=========

We're deliberately adding a bit of testing code to the production bundle (a
check whether the request to the stream returns plain JSON, which we want for
integration-testing). I did not find a satisfactory workaround in a reasonable time. Let's hope for a second pair of eyes or some #showerthoughts. :)

Also we're introducing a workaround to make cypress recognize and potentially
stub fetch-calls, as it is not capable of doing so on its own yet, see
cypress-io/cypress#95.

Closes DCOS-61250, DCOS-61253

### CODE REVIEW

File: src/js/core/MesosStream.ts [hunk 1]
```diff
@@ -1,11 +1,90 @@
...
+    Accept: "application/x-protobuf",
+    "Content-Type": "application/json"
+  }
+}).pipe(
+  switchMap(response => {
+    if (!response) {
+      throw Error("No response object returned for MesosStateStream");
+    }
+
+    // we only ever expect JSON to be served during integration-tests where
```
[the remark this answers is not in the document]
Comment (Poltergeist) [reply src/js/core/MesosStream.ts]: Cypress is not as bad as it sounds in other projects we use it very effectively maybe we want to get some learnings from there and solve the problems we have in the cypress setup in dcos-ui. And let's please not forget that moving to another testing system also means that we need to rewrite a lot of tests. Let's keep on discussing this offline since it is not part of this PR.

[the remark this answers is not in the document]
Comment (orlandohohmeier) [reply src/js/core/MesosStream.ts]: Well, it neither supports dynamic stubbing, modern browser APIs, nor multi browser testing and if we need to roll all of this our own than I don't see the value it adds. That being said I want us to consider dropping us that doesn't mean that we're going to as their may things it offers that are reason enough to keep it around.

File: src/js/core/MesosStream.ts [hunk 2]
```diff
@@ -1,11 +1,90 @@
...
+  return new Observable(observer => {
+    // a buffer to temporarily hold binary data that is streamed. it might be
+    // that mesos streams less than a full chunk - that's why we need to store
+    // that temporarily.
+    let buffer = new Uint8Array();
+
+    // @ts-ignore
+    reader.read().then(function process({ done, value }) {
+      if (done) {
+        return observer.complete();
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/core/MesosStream.ts]: totes magotes!

File: src/js/mesos-proto.js [hunk 3]
```diff
@@ -0,0 +1,2622 @@
+/*
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/mesos-proto.js]: i think a loader would be overengineered, as the mesos-protos seem to change very often.

what do you think of adding the current state of the protos to this repo along with a build-task?

File: src/js/core/MesosStream.ts [hunk 4]
```diff
@@ -1,11 +1,90 @@
+import { Observable, of, timer } from "rxjs";
+import { fromFetch } from "rxjs/fetch";
+import { map, repeatWhen, shareReplay, switchMap } from "rxjs/operators";
 // @ts-ignore
-import { stream } from "@dcos/mesos-client";
-import { timer } from "rxjs";
-import { repeatWhen } from "rxjs/operators";
+import proto from "../mesos-proto";
+
+const eventCoder = proto.lookup("Event");
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/core/MesosStream.ts]: i like! that's rather a historical name i did not revisit :)

File: src/js/core/MesosStream.ts [hunk 5]
```diff
@@ -1,11 +1,90 @@
...
+
+const utf8Coder = new TextDecoder("utf-8");
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/core/MesosStream.ts]: thinking about it, i actually never left the prototyping-mode as the integration-tests still don't seem to run. thus the polishing-mode was not activated yet.

it surely is a `decoder`. but right now i'm wondering whether we can do without `TextDecoder` at all. i'll give that a try and rename in case that idea does not work out.

File: src/js/core/MesosStream.ts [hunk 6]
```diff
@@ -1,11 +1,90 @@
...
+
+      // append the data that's coming in to our buffer
+      const newBuffer = new Uint8Array(buffer.length + value.length);
+      newBuffer.set(buffer);
+      newBuffer.set(value, buffer.length);
+      buffer = newBuffer;
+
+      // check if we have enough data in the buffer to emit a chunk
+      const newlineIndex = buffer.indexOf(10) + 1;
+      const chunkLengthStr = utf8Coder.decode(buffer.slice(0, newlineIndex));
```
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/core/MesosStream.ts]: 😍 totally!

File: src/js/core/MesosStream.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/core/MesosStream.ts]: i think i like the idea. i'll have a look at what Julian once put together to get a feeling for what that means in daily development.

File: src/js/mesos-proto.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (orlandohohmeier) [reply src/js/mesos-proto.js]: Works for me, also allows for the creation of type definition along with the _bindings_.

[the remark this answers is not in the document]
Comment (pierrebeitz) [reply src/js/mesos-proto.js]: well, the type definitions won't help us very much as the out-type for stuff form `toObject` (which we currently need to rely on) is `Record<string, any>`! 😢

i'll add the former stuff anyway. and if you think it's worthwhile to have (not very good IMHO) typings lying around somewhere in case we need them, i can add those as well, see https://github.com/dcos/dcos-ui/pull/4188.
